core/src/main/java/org/apache/accumulo/core/file/FileOperations.java

```diff
@@ -234,283 +207,314 @@ public FileSystem getFileSystem() {
-    /** Return the size of the file. */
-    long execute() throws IOException;
+    public BlockCache getDataCache() {
+      return dataCache;
+    }
+
+    public BlockCache getIndexCache() {
+      return indexCache;
+    }
+
+    public Cache<String,Long> getFileLenCache() {
+      return fileLenCache;
+    }
+
+    public boolean isSeekToBeginning() {
+      return seekToBeginning;
+    }
+
+    public CryptoService getCryptoService() {
+      return cryptoService;
+    }
+
+    public Range getRange() {
+      return range;
+    }
+
+    public Set<ByteSequence> getColumnFamilies() {
+      return columnFamilies;
+    }
+
+    public boolean isRangeInclusive() {
+      return inclusive;
+    }
   }
 
   /**
-   * Options common to all {@code FileOperation}s which perform reading or writing.
+   * Helper class extended by both writers and readers.
    */
-  protected static class FileIOOperation<SubclassType extends FileIOOperation<SubclassType>>
-      extends FileAccessOperation<SubclassType> {
-    private RateLimiter rateLimiter;
+  public class FileHelper {
+    protected AccumuloConfiguration tableConfiguration;
+    protected String filename;
+    protected FileSystem fs;
+    protected Configuration fsConf;
+    protected RateLimiter rateLimiter;
+
+    protected FileHelper fs(FileSystem fs) {
+      Objects.requireNonNull(fs);
+      this.fs = fs;
+      return this;
+    }
+
+    protected FileHelper fsConf(Configuration fsConf) {
+      Objects.requireNonNull(fsConf);
+      this.fsConf = fsConf;
+      return this;
+    }
 
-    /** Specify a rate limiter for this operation. */
-    @SuppressWarnings("unchecked")
-    public SubclassType withRateLimiter(RateLimiter rateLimiter) {
+    protected FileHelper filename(String filename) {
+      Objects.requireNonNull(filename);
+      this.filename = filename;
+      return this;
+    }
+
+    protected FileHelper tableConfiguration(AccumuloConfiguration tableConfiguration) {
+      Objects.requireNonNull(tableConfiguration);
+      this.tableConfiguration = tableConfiguration;
+      return this;
+    }
+
+    protected FileHelper rateLimiter(RateLimiter rateLimiter) {
       this.rateLimiter = rateLimiter;
-      return (SubclassType) this;
+      return this;
     }
 
-    public RateLimiter getRateLimiter() {
-      return rateLimiter;
+    protected FileOptions toWriterBuilderOptions(String compression,
+        FSDataOutputStream outputStream, boolean startEnabled) {
+      return new FileOptions(tableConfiguration, filename, fs, fsConf, rateLimiter, compression,
+          outputStream, startEnabled, null, null, null, false, null, null, null, true);
+    }
+
+    protected FileOptions toReaderBuilderOptions(BlockCache dataCache, BlockCache indexCache,
+        Cache<String,Long> fileLenCache, boolean seekToBeginning, CryptoService cryptoService) {
+      return new FileOptions(tableConfiguration, filename, fs, fsConf, rateLimiter, null, null,
```

Comment:
I spent some time checking these arguments and everything looks good. This comment is info about what I did.    I noticed that most of the consecutive arguments have different types, so that helps ensure things are in the correct order except for nulls.  The only two consecutive types are the cache's and those are in correct order (data 1st and index 2nd).  For the nulls, I verified that all of the fields of helper instance vars are passed and all of the arguments for each method are passed.  The three booleans were another possible area for bug, but those also seem good.

Reply:
Awesome thanks for checking that!